src/coreclr/jit/codegenxarch.cpp

```diff
@@ -1106,6 +1106,19 @@ void CodeGen::genCodeForMul(GenTreeOp* treeNode)
 
             inst_RV_SH(INS_shl, size, targetReg, shiftAmount);
         }
+#ifdef TARGET_AMD64
+        else if (!requiresOverflowCheck && rmOp->isUsedFromReg() && isPow2(imm + 1))
```

Comment:
I do think this is better here in codegen or maybe even in lowering.

That being said, the fact that it couldn't be done in morph because it introduced regressions particularly around `LEA` is likely a case we should log an issue around.

There are many devs who explicitly use `SHL`/`SHR`/`SAR` as an optimization rather than using a multiply and we should likely recognize that they are "equivalent" (and therefore can be CSE'd the same, can participate in containment the same, etc).